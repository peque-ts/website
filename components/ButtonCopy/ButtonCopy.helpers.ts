import { useEffect, useRef, useState } from 'react';

export function useClipboard(): [(data: string) => Promise<void>, boolean] {
  const [copied, setCopied] = useState(false);

  const timeout = useRef<number>();

  useEffect(() => {
    return () => {
      window.clearTimeout(timeout.current);
    };
  }, []);

  const copy = async (data: string) => {
    if (copied) {
      return;
    }

    try {
      await navigator.clipboard.writeText(data);
      setCopied(true);

      window.clearTimeout(timeout.current);
      timeout.current = window.setTimeout(() => {
        setCopied(false);
      }, 2000);
    } catch (error) {
      console.error('Unable to copy to clipboard.', error);
    }
  };

  return [copy, copied];
}
