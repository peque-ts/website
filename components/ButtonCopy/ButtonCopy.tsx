import React from 'react';

import { Icon } from '../Icon';
import { useClipboard } from './ButtonCopy.helpers';

interface Props {
  data: string;
}

export const ButtonCopy: React.VFC<Props> = ({ data }) => {
  const [copy, copied] = useClipboard();

  return (
    <button type="button" className="flex items-center" onClick={() => copy(data)}>
      {copied ? (
        <span className="text-sm">copied!</span>
      ) : (
        <>
          <span className="sr-only">Copy to clipboard</span>
          <Icon name="Copy" className="text-secondary-200 hover:text-white transition" size={16} />
        </>
      )}
    </button>
  );
};
