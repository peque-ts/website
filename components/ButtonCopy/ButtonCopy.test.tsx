import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ButtonCopy } from './ButtonCopy';

describe('ButtonCopy', () => {
  it('should render', () => {
    const { asFragment } = render(<ButtonCopy data="data to copy" />);

    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should copy data when clicked and browser supports it', async () => {
    const writeText = jest.fn();
    (global.navigator as any).clipboard = { writeText };

    render(<ButtonCopy data="data to copy" />);

    userEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(writeText).toHaveBeenCalledWith('data to copy');
    });

    (global.navigator as any).clipboard = undefined;
  });

  it('should log an error when clipboard is not supported', async () => {
    const spy = jest.spyOn(console, 'error').mockImplementation();

    render(<ButtonCopy data="data to copy" />);

    userEvent.click(screen.getByRole('button'));

    expect(spy).toHaveBeenCalledWith('Unable to copy to clipboard.', expect.any(Error));

    spy.mockRestore();
  });

  it('should reset its state after the timeout', async () => {
    jest.useFakeTimers();

    const writeText = jest.fn();
    (global.navigator as any).clipboard = { writeText };

    render(<ButtonCopy data="data to copy" />);

    userEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(writeText).toHaveBeenCalledTimes(1);
    });

    expect(screen.getByText('copied!')).toBeInTheDocument();

    userEvent.click(screen.getByRole('button'));

    expect(writeText).not.toHaveBeenCalledTimes(2);

    act(() => {
      jest.advanceTimersByTime(4000);
    });

    expect(screen.queryByText('copied!')).toBeNull();

    (global.navigator as any).clipboard = undefined;
    jest.useRealTimers();
  });
});
