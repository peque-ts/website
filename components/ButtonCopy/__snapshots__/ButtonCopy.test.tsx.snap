// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonCopy should render 1`] = `
<DocumentFragment>
  <button
    class="flex items-center"
    type="button"
  >
    <span
      class="sr-only"
    >
      Copy to clipboard
    </span>
    <svg
      class="text-secondary-200 hover:text-white transition"
      height="16"
      viewBox="0 0 64 64"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        transform="rotate(.045) scale(4.57143)"
      >
        <rect
          fill="none"
          height="10.5"
          rx="1"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          width="10.5"
          x=".5"
          y=".5"
        />
        <path
          d="M13.5 3.5v9a1 1 0 0 1-1 1h-9"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </g>
    </svg>
  </button>
</DocumentFragment>
`;
